### PR TITLE
Fix CORS Support

### PR DESCRIPTION
**Pull request by jbenet**
This PR
- Fixes CORS bugs
- Improves support for CORS options
- Removes Referrer Checking
- Defaults to localhost/127.0.0.1 CORS Origins
- Allows Setting arbitrary HTTP headers on API responses

### CODE REVIEW

**Comment by jbenet** [comment]
@whyrusleeping @cryptix PTAL

File: commands/http/handler.go [hunk 1]
```diff
@@ -46,33 +47,61 @@ const (
...
+	// Headers is an optional map of headers that is written out.
+	Headers map[string][]string
+
+	// CORSOpts is a set of options for CORS headers.
+	CORSOpts *cors.Options
+}
+
+func skipAPIHeader(h string) bool {
+	switch h {
+	default:
```
Comment (whyrusleeping) [inline commands/http/handler.go]: seeing default first and implicit continues always weirds me out in the logic. maybe:
```
switch h{
    case "Access-Control-Allow-Origin":
       return true
    case "Access-Control-Allow-Methods":
       return true
    case "Access-Control-Allow-Credentials":
       return true
    default:
        return false
}
```

just to be explicit?

File: commands/http/handler.go [hunk 2]
```diff
@@ -153,7 +175,7 @@ func guessMimeType(res cmds.Response) (string, error) {
 	return mimeTypes[enc], nil
 }
 
-func sendResponse(w http.ResponseWriter, req cmds.Request, res cmds.Response) {
+func sendResponse(w http.ResponseWriter, r *http.Request, req cmds.Request, res cmds.Response) {
```
Comment (whyrusleeping) [inline commands/http/handler.go]: the ordering here feels weird. maybe switch res and req to stay consistent? http{resp, req}, cmds{resp, req}

**Comment by whyrusleeping** [comment]
LGTM

**Comment by mappum** [comment]
CORS does not prevent CSRF attacks. Here is an attack example, if your daemon is on this branch then visiting the page will pin something to your node: http://localhost:8080/ipfs/QmYeRGr8PY3yigx2aTvK8B9e4HVUy3h4b2r3tsZHaYBduD

(Note that the attack can come from any website you visit, it doesn't have to be an IPFS-hosted page).

Pinning arbitrary things is only a sort-of scary attack, but more critical attacks could be possible too (e.g. making someone publish the attacker's hash for their IPNS names).

**Comment by jbenet** [comment]
@mappum `QmYeRGr8PY3yigx2aTvK8B9e4HVUy3h4b2r3tsZHaYBduD` is not resolving -- pin on irc?

**Comment by diasdavid** [comment]
@mappum To prevent CSRF, we should have some kind of token (a CSRF token of a JSON WebToken), blocking by referer header doesn't make it impenetrable and it can be annoying for web app devs.

**Comment by jbenet** [comment]
+1 for token, though it can also be annoying in its own right.

maybe:
```
var ipfs = ipfsApi(...)
```
could get the token async? not sure.


@mappum and @diasdavid would you discuss and decide? you know more about all the constraints than me

**Comment by jbenet** [comment]
I would really like to fix + merge this so we can release 0.3.6 soon-- it has _lots_ of important bugfixes.

**Comment by jbenet** [comment]
aaaand done.

**Pull request merged**